**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Carlos Joel Bugarín González
*   MATRÍCULA: A01841627

---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [ ]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [ ]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [ ]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# Los datos de conexión (usuario, clave y host) están en la celda original de esta libreta, en Canvas o con tu profesora.
# Se piden con getpass para que no queden escritos en el código ni en las salidas que se publican en GitHub.
import getpass

usuario = getpass.getpass('Usuario de la BD: ')
clave = getpass.getpass('Clave de la BD: ')
host = getpass.getpass('Host (IP) del servidor: ')

# URL.create escapa automáticamente los caracteres especiales de la clave (por ejemplo ! o @)
url_conexion = sqla.engine.URL.create(
    drivername='mysql+pymysql',
    username=usuario,
    password=clave,
    host=host,
    port=3306,
    database='Netflix',
)

# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine(url_conexion, pool_recycle=3600)
conn = db.connect()

# Se eliminan de la memoria las variables sensibles
del usuario, clave, host, url_conexion

Usuario de la BD: ··········
Clave de la BD: ··········
Host (IP) del servidor: ··········


Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [ ]:
# runtime está en minutos: 5 horas = 300 minutos
query = sqla.text("""
    SELECT *
    FROM movie
    WHERE runtime > 300
""")
pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [ ]:
# COUNT(columna) ignora los NULL; SUM(available_globally = 0) cuenta las películas disponibles solo en EU
query = sqla.text("""
    SELECT
        ROUND(100.0 * SUM(available_globally = 0) / COUNT(available_globally), 2)
            AS porcentaje_solo_eu
    FROM movie
""")
pd.read_sql(query, conn)

,porcentaje_solo_eu
0,83.57


3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [ ]:
# Valores distintos de locale (si existe NULL, aparece como None en pandas)
query = sqla.text("""
    SELECT DISTINCT locale
    FROM movie
    ORDER BY locale
""")
pd.read_sql(query, conn)

,locale
0,None
1,en


In [ ]:
# Para comparar con NULL en SQL se usa IS NULL (nunca = NULL)
query = sqla.text("""
    SELECT COUNT(*) AS registros_locale_nulo
    FROM movie
    WHERE locale IS NULL
""")
pd.read_sql(query, conn)

,registros_locale_nulo
0,11255


**Análisis:**

La tabla contiene dos valores distintos de `locale`: `en` y NULL. Hay 11,255 películas con `locale` NULL, que representan el 95.49 % de las 11,787 registradas. Solo 532 películas (4.51 %) tienen `locale` informado, y todas son `en`. Esto indica que el campo está vacío en casi toda la tabla, por lo que no permite identificar el idioma original de la mayoría de las películas.

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [ ]:
# (a) Películas cuyo título difiere del título original
query = sqla.text("""
    SELECT COUNT(*) AS peliculas_titulos_distintos
    FROM movie
    WHERE title <> original_title
""")
pd.read_sql(query, conn)

,peliculas_titulos_distintos
0,3947


In [ ]:
# (b) Tabla cruzada: locale nulo (1 = sí, 0 = no) contra títulos distintos (1 = sí, 0 = no)
# Si title u original_title fueran NULL, la comparación da NULL y aparecería como un grupo aparte
query = sqla.text("""
    SELECT
        (locale IS NULL)          AS locale_nulo,
        (title <> original_title) AS titulos_distintos,
        COUNT(*)                  AS registros
    FROM movie
    GROUP BY locale_nulo, titulos_distintos
    ORDER BY locale_nulo, titulos_distintos
""")
pd.read_sql(query, conn)

,locale_nulo,titulos_distintos,registros
0,0,NaN,532
1,1,NaN,7291
2,1,0.0,17
3,1,1.0,3947


In [ ]:
# (c1) Registros con locale informado pero título distinto: ¿qué locales son?
query = sqla.text("""
    SELECT locale, COUNT(*) AS registros
    FROM movie
    WHERE locale IS NOT NULL
      AND title <> original_title
    GROUP BY locale
    ORDER BY registros DESC
""")
pd.read_sql(query, conn)

,locale,registros


In [ ]:
# (c2) Muestra de registros con locale nulo pero título idéntico al original
query = sqla.text("""
    SELECT id, title, original_title, available_globally, release_date
    FROM movie
    WHERE locale IS NULL
      AND title = original_title
    LIMIT 10
""")
pd.read_sql(query, conn)

,id,title,original_title,available_globally,release_date
0,1089,Recep Ivedik 7,Recep İvedik 7,b'\x00',None
1,1349,Veronica,Verónica,b'\x00',None
2,3509,Fantomas,Fantômas,b'\x00',None
3,3611,Luccas Neto em: O Hotel Magico,Luccas Neto em: O Hotel Mágico,b'\x00',None
4,4566,Un Dia En El Paraiso,Un día en el paraíso,b'\x00',None
5,5077,Superlopez,Superlópez,b'\x01',None
6,5211,Baris Akarsu Merhaba,Barış Akarsu Merhaba,b'\x00',None
7,5684,Totem,Tótem,b'\x00',None
8,6427,La Pena Maxima,La pena máxima,b'\x01',None
9,6530,Mirciulica,Mirciulică,b'\x00',None


**Análisis:**

- Hay 3,947 películas con `title` distinto de `original_title`, frente a 11,255 con `locale` NULL. Los resultados no coinciden.
- De las 11,255 películas con `locale` NULL: 3,947 (35.07 %) tienen títulos distintos; 7,291 (64.78 %) no se pueden comparar porque al menos uno de los dos títulos es NULL; y 17 (0.15 %) aparecen con títulos "iguales".
- Esas 17 difieren en acentos o mayúsculas (por ejemplo "Veronica" y "Verónica"), diferencias que MySQL ignora al comparar. Son títulos de películas en español, portugués, turco y rumano, que no están en inglés.
- Las 532 películas con `locale` `en` tienen al menos un título NULL, y ninguna muestra títulos distintos.
- **Conclusión:** la suposición se sostiene solo parcialmente. En todas las películas con `locale` NULL que se pueden comparar (3,964), los títulos difieren o difieren en acentos, lo que es consistente con un idioma distinto del inglés. Pero el 64.78 % de las películas con `locale` NULL carece de datos para comprobarlo, y que el 95.49 % de la tabla tenga `locale` NULL sugiere datos faltantes más que idioma distinto. Además, un título distinto puede deberse a traducciones o transliteraciones. Por eso `locale` NULL no basta como indicador confiable del idioma.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [ ]:
# cumulative_weeks_in_top_10 es un acumulado por periodo: su máximo por película son las semanas totales en el top 10
# El INNER JOIN con movie descarta las filas de series (movie_id NULL) y permite obtener el título
query = sqla.text("""
    SELECT
        m.title,
        MAX(vs.cumulative_weeks_in_top10) AS semanas_en_top_10
    FROM view_summary AS vs
    INNER JOIN movie AS m
        ON vs.movie_id = m.id
    GROUP BY m.id, m.title
    ORDER BY semanas_en_top_10 DESC
    LIMIT 1
""")
pd.read_sql(query, conn)

,title,semanas_en_top_10
0,The Boss Baby,22


6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [ ]:
# Cada fila de season es una temporada; se cuentan por serie y se toman las 5 con más temporadas
query = sqla.text("""
    SELECT
        tv.title,
        COUNT(se.id) AS numero_de_temporadas
    FROM tv_show AS tv
    INNER JOIN season AS se
        ON se.tv_show_id = tv.id
    GROUP BY tv.id, tv.title
    ORDER BY numero_de_temporadas DESC
    LIMIT 5
""")
pd.read_sql(query, conn)

,title,numero_de_temporadas
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [ ]:
# Primero se consulta qué valores toma duration, para no suponer el texto exacto
query = sqla.text("""
    SELECT DISTINCT duration
    FROM view_summary
""")
pd.read_sql(query, conn)

,duration
0,SEMI_ANNUALLY
1,WEEKLY


In [ ]:
# Valor de duration que corresponde a los periodos semestrales (según la celda anterior)
DURACION_SEMESTRAL = 'SEMI_ANNUALLY'

# :duracion es un parámetro enlazado; su valor se envía aparte en params
query = sqla.text("""
    SELECT DISTINCT start_date, end_date
    FROM view_summary
    WHERE duration = :duracion
    ORDER BY start_date
""")
pd.read_sql(query, conn, params={'duracion': DURACION_SEMESTRAL})

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [ ]:
# Ruta de relaciones: tv_show -> season (tv_show_id) -> view_summary (season_id)
# En LIKE, _ representa un carácter cualquiera: 'Grey_s Anatomy' coincide con el apóstrofe de Grey's
# El primer semestre de 2024 se identifica con duration semestral y start_date = 2024-01-01 (ver pregunta 7)
query = sqla.text("""
    SELECT
        se.season_number,
        se.title AS temporada,
        vs.views,
        vs.hours_viewed
    FROM tv_show AS tv
    INNER JOIN season AS se
        ON se.tv_show_id = tv.id
    INNER JOIN view_summary AS vs
        ON vs.season_id = se.id
    WHERE tv.title LIKE :patron
      AND vs.duration = :duracion
      AND vs.start_date = :inicio
    ORDER BY vs.views DESC, vs.hours_viewed DESC, se.season_number
""")
resultado_p8 = pd.read_sql(
    query,
    conn,
    params={'patron': 'Grey_s Anatomy', 'duracion': DURACION_SEMESTRAL, 'inicio': '2024-01-01'},
)
resultado_p8

,season_number,temporada,views,hours_viewed
0,1,Grey's Anatomy: Season 1,3600000,23400000
1,2,Grey's Anatomy: Season 2,3100000,60600000
2,3,Grey's Anatomy: Season 3,2900000,53700000
3,5,Grey's Anatomy: Season 5,2900000,49300000
4,4,Grey's Anatomy: Season 4,2900000,35300000
5,6,Grey's Anatomy: Season 6,2800000,48000000
6,7,Grey's Anatomy: Season 7,2700000,43000000
7,8,Grey's Anatomy: Season 8,2600000,44700000
8,9,Grey's Anatomy: Season 9,2500000,43300000
9,10,Grey's Anatomy: Season 10,2400000,41700000


**Respuesta:** en el primer semestre de 2024 (del 2024-01-01 al 2024-06-30), las temporadas de *Grey's Anatomy* ordenadas de mayor a menor número de vistas son: 1, 2, 3, 5, 4, 6, 7, 8, 9, 10, 11, 12, 13, 14, 19, 15, 16, 18, 17 y 20. Los empates en vistas se desempatan por horas vistas y, después, por número de temporada.

**Análisis:**

- **Extremos:** la temporada 1 tiene más vistas (3,600,000) y la temporada 20 tiene menos (100,000).
- **Patrón:** las vistas descienden de forma gradual con el número de temporada: 3.6 M en la temporada 1, 3.1 M en la 2, unos 2.9 M en las temporadas 3 a 5, entre 2.8 M y 2.4 M en las 6 a 10, entre 2.2 M y 2.0 M en las 11 a 14, y entre 1.9 M y 1.7 M en las 15 a 18. La excepción es la temporada 19 (2.0 M), que supera a las temporadas 15 a 18. La temporada 20 queda muy por debajo del resto.
- **Interpretación:** el patrón es consistente con un embudo de audiencia: la primera temporada funciona como puerta de entrada y, en cada temporada posterior, parte de quienes empezaron la serie dejan de verla. Que la temporada 1 lidere indica que el interés se concentra en el inicio de la serie.
- **Vistas frente a horas vistas:** con `hours_viewed` el orden sería distinto: la temporada 2 tendría más horas (60.6 M) que la temporada 1 (23.4 M). La diferencia se explica por la duración de cada temporada: según la definición de Netflix, `views` son las horas vistas divididas entre la duración del título, lo que da aproximadamente 6.5 horas para la temporada 1 y 19.6 horas para la 2. Por eso `views`, que normaliza por duración, refleja mejor cuánta gente vio cada temporada.
- **Casos que requieren cautela:** la temporada 20 podría tener pocas vistas por haber estado disponible solo una parte del periodo, y la temporada 19 podría reflejar un repunte de interés. Los datos no permiten confirmar ninguna de las dos causas; son hipótesis a verificar.
- **Limitaciones:** el análisis cubre un solo semestre y solo la audiencia dentro de Netflix, y las cifras aparecen redondeadas a cientos de miles, lo que genera empates y resta precisión.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

**Respuesta:** la consulta de la pregunta 8 se replicó con pandas: se filtró la serie *Grey's Anatomy* en `tv_show`, se unieron sus temporadas con `merge` por `tv_show_id`, se unieron las vistas del primer semestre de 2024 con `merge` por `season_id` y se ordenó por `views`, `hours_viewed` y `season_number`. El resultado es el mismo que el de la consulta SQL: las mismas 20 temporadas, con los mismos valores y el mismo orden. La única diferencia es de presentación: pandas muestra `season_number` como número decimal.

In [ ]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [ ]:
# 1) Serie: equivalente a WHERE tv.title LIKE 'Grey_s Anatomy' (en la expresión regular, el punto equivale a "_")
serie = tv_show.loc[
    tv_show['title'].str.fullmatch(r'Grey.s Anatomy', case=False, na=False),
    ['id'],
].rename(columns={'id': 'tv_show_id'})

# 2) Temporadas: se renombran columnas para que el merge sea claro y sin columnas duplicadas
temporadas = season[['id', 'tv_show_id', 'season_number', 'title']].rename(
    columns={'id': 'season_id', 'title': 'temporada'}
)

# 3) Periodo: duration semestral y start_date = 2024-01-01 (se convierte a fecha para comparar)
periodo = view_summary.loc[
    (view_summary['duration'] == DURACION_SEMESTRAL)
    & (pd.to_datetime(view_summary['start_date']) == pd.Timestamp('2024-01-01')),
    ['season_id', 'views', 'hours_viewed'],
].dropna(subset=['season_id'])
periodo = periodo.astype({'season_id': 'int64'})  # season_id llega como decimal por los NULL

# 4) Equivalente a los dos INNER JOIN + ORDER BY views DESC
resultado_p9 = (
    serie
    .merge(temporadas, on='tv_show_id', how='inner')
    .merge(periodo, on='season_id', how='inner')
    .sort_values(['views', 'hours_viewed', 'season_number'], ascending=[False, False, True])
    .loc[:, ['season_number', 'temporada', 'views', 'hours_viewed']]
    .reset_index(drop=True)
)
resultado_p9

,season_number,temporada,views,hours_viewed
0,1.0,Grey's Anatomy: Season 1,3600000,23400000
1,2.0,Grey's Anatomy: Season 2,3100000,60600000
2,3.0,Grey's Anatomy: Season 3,2900000,53700000
3,5.0,Grey's Anatomy: Season 5,2900000,49300000
4,4.0,Grey's Anatomy: Season 4,2900000,35300000
5,6.0,Grey's Anatomy: Season 6,2800000,48000000
6,7.0,Grey's Anatomy: Season 7,2700000,43000000
7,8.0,Grey's Anatomy: Season 8,2600000,44700000
8,9.0,Grey's Anatomy: Season 9,2500000,43300000
9,10.0,Grey's Anatomy: Season 10,2400000,41700000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [ ]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/drive/MyDrive/Colab Notebooks MNA TC5053 - Ciencia y analítica de datos Actividad3_BD_Manipulacion"
os.chdir(DIR)

In [ ]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [ ]:
# La conexión a Firestore (db) se creó en la celda de autenticación
# Cada documento se convierte en un diccionario que incluye su identificador (doc.id)
documentos = db.collection('EMPLOYEE').stream()
empleados = [{'id': doc.id, **doc.to_dict()} for doc in documentos]

# El id se usa como índice y se ordenan las columnas como en el ejemplo del enunciado
df_empleados = (
    pd.DataFrame(empleados)
    .set_index('id')[['emp_fname', 'emp_lname', 'emp_hiredate']]
)
df_empleados

,emp_fname,emp_lname,emp_hiredate
id,,,
8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


**Respuesta:** la colección `EMPLOYEE` de Firestore contiene 4 documentos. Cada uno se leyó con `stream()`, se convirtió en un diccionario que incluye su identificador (`doc.id`) y se cargó en un dataframe de pandas que usa el `id` como índice, con las columnas `emp_fname`, `emp_lname` y `emp_hiredate`.

In [ ]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA y Declaración del autor:**

*   Anthropic. (2026). *Claude (Claude Sonnet 5.5)* [Modelo de lenguaje grande], utilizado para validación y modificación de código propio (consultas SQL, código de pandas y lectura de Firestore), depuración de errores de ejecución. https://claude.ai/

El código fue revisado, ejecutado, validado por el autor con los resultados obtenidos en esta libreta y se agregaron las respuestas correspondientes; las cifras del análisis provienen de las salidas del propio servidor.

---